datalad/core/local/resulthooks.py

```diff
@@ -0,0 +1,174 @@
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Interface utility functions
+
+"""
+
+__docformat__ = 'restructuredtext'
+
+import logging
+import json
+
+lgr = logging.getLogger('datalad.core.local.hooks')
+
+
+def get_hooks_from_config(cfg):
+    """Parse out hook definitions given a ConfigManager instance
+
+    Returns
+    -------
+    dict
+      where keys are hook names/labels, and each value is a dict with
+      three keys: 'cmd' contains the name of the to-be-executed DataLad
+      command; 'args' has a JSON-encoded string with a dict of keyword
+      arguments for the command (format()-language based placeholders
+      can be present; 'match' hold a JSON-encoded string representing
+      a dict with key/value pairs that need to match a result in order
+      for a hook to be triggered.
+    """
+    hooks = {}
+    for h in (k for k in cfg.keys()
+              if k.startswith('datalad.result-hook.')
+              and k.endswith('.match')):
+        proc = cfg.get('{}.proc'.format(h[:-6]), None)
+        if not proc:
+            lgr.warning(
+                'Incomplete result hook configuration %s in %s' % (
+                    h[:-6], cfg))
+            continue
+        sep = proc.index(' ')
+        hooks[h[20:-6]] = dict(
+            cmd=proc[:sep],
+            args=proc[sep + 1:],
+            match=json.loads(cfg.get(h)),
+        )
+    return hooks
+
+
+def match_hook2result(hook, res, match):
+    """Evaluate a hook's result match definition against a concrete result
+
+    A match definition is a dict that can contain any number of keys. For each
+    key it is tested, if the value matches the one in a given result.
+    If all present key/value pairs match, the hook is executed. In addition to
+    ``==`` tests, ``in``, ``not in``, and ``!=`` tests are supported. The
+    test operation can be given by wrapping the test value into a list, the
+    first item is the operation label 'eq', 'neq', 'in', 'nin'; the second value
+    is the test value (set). Example::
+
+        {
+          "type": ["in", ["file", "directory"]],
+          "action": "get",
+          "status": "notneeded"
+        }
+
+    Parameters
+    ----------
+    hook : str
+      Name of the hook
+    res : dict
+      Result dictionary
+    match : dict
+      Match definition (see above for details).
+
+    Returns
+    -------
+    bool
+      True if the given result matches the hook's match definition, or
+      False otherwise.
+    """
+    for k, v in match.items():
+        # do not test 'k not in res', because we could have a match that
+        # wants to make sure that a particular value is not present, and
+        # not having the key would be OK in that case
+
+        # in case the target value is an actual list, an explicit action 'eq'
+        # must be given
```

Comment:
This needs to be mentioned in the docstring.

Reply:
True

Reply:
Done